### PR TITLE
Add BigQuery UDF support

### PR DESCRIPTION
**Pull request by takbab**
Add a field to support BigQuery UDF.
https://cloud.google.com/bigquery/docs/reference/v2/jobs
- configuration.query.userDefinedFunctionResources[]
- configuration.query.userDefinedFunctionResources[].inlineCode
- configuration.query.userDefinedFunctionResources[].resourceUri

(migrated from #1988)

### CODE REVIEW

File: gcloud/bigquery/job.py [hunk 1]
```diff
@@ -937,6 +938,12 @@ def __init__(self, name, query, client):
     reference/v2/jobs#configuration.query.useLegacySql
     """
 
+    udf_resources = _TypedProperty(
```
Comment (takbab) [inline gcloud/bigquery/job.py]: commented by @tseaver 
https://github.com/GoogleCloudPlatform/gcloud-python/pull/1988/files/b596c49220945f854c432a567ce08f49eafcc219#r71068506
>Using _TypedProperty for a list of mappings leaves us with no checking at assignment time. @dhermes, do you think we should have a "setter" for the property which checks that the list elements are dicts with appropriate keys?

Comment (dhermes) [inline gcloud/bigquery/job.py]: @tseaver This should probably have a more dedicate property class in place

**Comment by tseaver** [comment]
@takbab I think @dhermes was talking about dropping the use of `_TypedProperty` in favor of a property implementation which enforces that items in the list are suitable.  If we use your class for the items in the list, then the property implementation would look something like:

```python

    @property
    def udf_resources(self):
        """List of user-defined resources.

        :rtype: list of :class:`UdfResource`
        :returns: the list, if set, or None
        """
        return list(self._udf_resources) if self._udf_resources else None

    @udf_resources.setter
    def udf_resources(self, value):
        """Update list of user-defined resources

        :type value: list of :class:`UdfResource`
        :param value: the new list

        :raises: ValueError, ir the items in the list are not all instances of
                 :class:`UdfResource`
        """
        if not all((isinstance(item, UdfResource) for item in value)):
            raise ValueError("pass a list of UdfResource instances")
        self._udf_resources = tuple(value)
```

**Comment by tseaver** [comment]
See #2015 for an alternative.

**Comment by dwmclary** [comment]
#2015 looks a more like the example from @tseaver but I need to clean up a lint error.  What's the advantage to the approach here?

**Comment by takbab** [comment]
@tseaver Thanks for the example! I have added the property to QueryJob class.

What can we do?
I just want to use UDF from my app.
#1829

**Comment by tseaver** [comment]
@takbab Thanks again for the patch!  I merged @dwmclary's version of the feature in #2029, as it looked closer to what I had in mind.

**Comment by takbab** [comment]
I see, thanks!

**Pull request closed without merging**